doc/developer/design/20230519_statement_logging.md

```diff
@@ -0,0 +1,165 @@
+        id: Uuid,
+        prepared_statement_id: Uuid,        
+        params: Vec<Datum>,
+        when: DateTime,
+        // Additional metadata might eventually
+        // go here, e.g. `plan: Option<OptimizedMirRelationExpr>`
+    },
+    StatementExecutionFinished {
+        id: Uuid,
+        when: DateTime,
+        n_results: Option<usize>,
+    },
+}
+```
+
+The `StatementPrepared` and `StatementExecutionBegan` events are
+generated and stored in the buffer in `fn handle_execute`, after
+checking a random value against `STATEMENT_LOG_SAMPLING_RATE` . This
+requires keeping track of the prepared statement bound to each name,
+and whether it has been logged yet (as we don't want to log each
+statement several times). The `StatementExecutionFinished` events are
+generated in several different places, depending on the type of
+statement. For example, the event for `SELECT` queries is generated in
+response to the `ControllerResponse::PeekResponse` message.
+
+The coordinator also holds on to an interval object whose tick rate is
+controlled by `STATEMENT_LOG_FLUSH_FREQUENCY`. It listens for ticks
+on that interval in its main `serve` loop; when such a tick occurs, it
+emits data to the system tables, as follows:
+
+* Each `StatementPrepared` event generates an entry in
+  `mz_prepared_statement_history`.
+* Each `StatementExecutionBegan` event generates an entry in
+  `mz_statement_execution_history`, and records that entry for
+  retraction as part of a later update.
+* Each `StatementExecutionEnded` event generates an update to the
+  entry described above.
+
+These entries are consolidated and then emitted to the builtin
+tables, and also to the catalog, ensuring durability.
+
+On system restart, all entries older than `STATEMENT_LOG_MAX_AGE` are
+deleted.
+
+### Observability and Rollout
+
+If we misjudge the size and volume of our typical users' queries, we
+will log much more data than we expect. Thus, rollout should proceed
+in two phases. First, we should enable the feature in "dry run" mode,
+which does not actually log any queries, but tracks (as a Prometheus
+metric) how much data would have been logged had the feature actually
+been turned on. Only once we validate that it meets expectations
+should we actually enable it. We should do something similar before
+adding any new columns to the tables (e.g., the plans).
+
+### Testing
+
+For testing, we will probably need a way to make the random sampling
+deterministic; this can be accomplished by adding a
+`STATEMENT_LOG_RANDOM_SEED` system var that we set from within
+testdrive or SLTs.
+
+## Alternatives
+
+Log all executed statements, rather than sampling. This approach was
+rejected because users might come to rely on all statements being logged,
+which would require us making all statements durable, dramatically
+increasing the cost of high-QPS usage patterns. Instead, we will not
+allow setting the sampling rate above some value slightly less than 1,
+e.g. 0.99.
```

Comment:
We may also want to note that it's easy to go from .99 to 1.0 in the future if we decide we can handle the load and consistency requirements (or are willing to sacrifice on the consistency), but the reverse is not true.